### ISSUE TITLE
use meshgrid for some components instead of broadcasting

### ISSUE THREAD

**Issue by xzackli** [simonsobs/fgspectra#16]
The numpy broadcasting is potentially confusing, might be good to have some examples which use `np.meshgrid` instead. Consider what I had to write for `tSZxCIB` for example,

```python
return (2 * (ThermalSZ.f(nu_1, T_CMB)[...,np.newaxis] * CIB.mu(nu_2, beta, T_d, T_CMB) +
                     CIB.mu(nu_1, beta, T_d, T_CMB)[...,np.newaxis] * ThermalSZ.f(nu_2, T_CMB))
/ (2 * ThermalSZ.f(nu_0, T_CMB) * CIB.mu(nu_0, beta, T_d, T_CMB)))
```

The `[..., np.newaxis]` is very numpy but kind of tricky to work with / understand.

**Comment by dpole** [comment]
I 100% agree that we need some examples. I'm not clear about how you would like to use `np.meshgrid`. Can you elaborate a bit on it?
About your tSZxCIB class, I understand you are in the `add_models_back_in` branch. What about opening a PR so that we can comment all the changes wrt master (otherwise we can comment only individual commits, as far as I know)?

**Comment by xzackli** [comment]
I've made the pull request! I had to debug some effective frequency discrepancies (ew) in order to make the code agree with the old Fortran stuff. 

Regarding `np.meshgrid`, here's what I was thinking. If we had some utility functions `get_ell)` and `get_freqs()` which returned arrays of ells and frequencies all in the shape `(freqs, freqs, ell)` but had those values, users could trivially define classes with code like

```python
def __call__(self, ... ):
    ell_mat = get_ell_mat(ells)
    f1_mat, f2_mat = get_freqs_mat(f1, f2)

    return (ell / 3000)**2 * (f1 * f2 / 150)**beta # this is shaped like (freqs, freqs, ells)
```

Maybe it doesn't need to be composite like this -- but this would help a lot with writing custom, nonfactorizable new components, for example?

**Comment by dpole** [comment]
I believe that this issue was addressed in the new definition of tSZxCIB (#19). More in general, as you say, the use of `meshgrid` is applicable only for non-factorizable components. I feel at ease with broadcasting, but I agree that the user is free of using `meshgrid` for new components.
I'm closing the issue, but feel free to reopen if you believe that there is anything else to discuss or decide about.
